File: 2025/JuyoungMoon/week1/2025-06-08-top-k-frequent-elements.cpp
// Author: Juyoung Moon
// Solved on Wed, June 4, 2025 (KST).

// KDD LeetCode Study Week 1: Array & Hashing
// https://github.com/juyomo/leetcode-study

// https://leetcode.com/problems/top-k-frequent-elements/

class Solution {
public:
    static bool compare(const vector<int>& a, const vector<int>& b) {
        // most frequent first
        return a[0] > b[0];
    }

    vector<int> topKFrequent(vector<int>& nums, int k) {
        // Keep track of how many of each num is in array.
        unordered_map<int, int> counts;
        for (int n : nums) {
            counts[n]++;
        }

        // Put that data into a vector.
        vector<vector<int>> countAndNum(counts.size(), vector<int>(2));
        int idx = 0;
        for (const auto& p : counts) {
            int num = p.first;
            int count = p.second;
            countAndNum[idx][0] = count;
            countAndNum[idx][1] = num;
            idx++;
        }

        // Sort from most frequent to least frequent. Meh, could do better time complexity.
        sort(countAndNum.begin(), countAndNum.end(), compare);

        // Return the top k frequent.
        vector<int> res(k);
        for (int i = 0; i < k; i++) {
            res[i] = countAndNum[i][1];
        }
        return res;
    }
};

File: 2025/Jaehyuk/2025-06-08-Top-K-Frequent-Elements.py
from collections import defaultdict


class Solution:
    def topKFrequent(self, nums: list[int], k: int) -> list[int]:
        ans = defaultdict(int)

        for num in nums:
            ans[num] += 1

        temp = []
        for num, cnt in ans.items():
            temp.append([cnt, num])
        temp.sort()

        res = []
        while len(res) < k:
            res.append(temp.pop()[1])

        return res
    
class Solution2: #better efficient solution
    def topKFrequent(self, nums: list[int], k: int) -> list[int]:
        count = dict()
        freq = [ [] for i in range(len(nums)+1)]

        for n in nums:
            count[n] = 1 + count.get(n, 0)
        
        for n, c in count.items():
            freq[c].append(n)

        res = []
        for i in range(len(freq) - 1, 0, -1):
            for n in freq[i]:
                res.append(n)
                if len(res) == k:
                    return res